README.md
[![Build Status](https://circleci.com/gh/jn-jairo/laravel-cast.svg?style=shield)](https://circleci.com/gh/jn-jairo/laravel-cast)
[![Total Downloads](https://poser.pugx.org/jn-jairo/laravel-cast/downloads)](https://packagist.org/packages/jn-jairo/laravel-cast)
[![Latest Stable Version](https://poser.pugx.org/jn-jairo/laravel-cast/v/stable)](https://packagist.org/packages/jn-jairo/laravel-cast)
[![License](https://poser.pugx.org/jn-jairo/laravel-cast/license)](https://packagist.org/packages/jn-jairo/laravel-cast)

# Cast for Laravel

This package provide cast for Laravel.

## Requirements

- Laravel Framework >= 5.8

## Installation

You can install the package via composer:

```bash
composer require jn-jairo/laravel-cast
```

The `CastServiceProvider` will be automatically registered for you.

## Usage

Both [contract](https://laravel.com/docs/contracts) `\JnJairo\Laravel\Cast\Contracts\Cast` and [facade](https://laravel.com/docs/facades) `\JnJairo\Laravel\Cast\Facades\Cast` are available.

There are three methods `Cast::cast()`, `Cast::castDb()` and `Cast::castJson()`.

- `Cast::cast()` casts to `PHP` types
- `Cast::castDb()` casts to `database` types
- `Cast::castJson()` casts to `json` types

All methods accept three parameters `mixed $value`, `string $type` and optionally `string $format`.

## Examples

```php
print_r(Cast::cast('{"foo":"bar"}', 'array'));
/*
Array
(
    [foo] => bar
)
*/

print_r(Cast::castDb(1234.555, 'decimal', '10:2'));
// 1234.56

print_r(Cast::castDb(['foo' => 'bar'], 'json'));
// {"foo":"bar"}

print_r(Cast::castJson(new DateTime('01 jan 2000'), 'date'));
// 2000-01-01
```

## Types available

- `int`, `integer`
- `float`, `real`, `double`
- `decimal`
- `bool`, `boolean`
- `date`
- `datetime`
- `timestamp`
- `json`
- `array`
- `object`
- `collection`
- `string`, `text`
- `uuid`
- `encrypted`
- `enum`

## Format parameter

- **decimal** - `precision:places|(up|down|ceiling|floor|half_up|half_down|half_even|half_odd|truncate)`. Example: `10:2|half_up`, `10:2`, `2`, `half_up`. Default: `28:2|half_up`.
The decimal type uses the https://php-decimal.io extension, to use this type run `composer require php-decimal/php-decimal:^1.1` and install the decimal extension.
- **date** - Example: `Y-m-d`. Default: `Y-m-d`.
- **datetime**, **timestamp** - Example: `Y-m-d H:i:s`. Default: `Y-m-d H:i:s`.
- **uuid** - `(uuid1|uuid4|ordered)`. Example: `uuid1`. Default: `uuid4`.
Empty string value will return a new UUID.
To use ordered UUID format run `composer require moontoast/math:^1.1`.
- **encrypted** - `type:format`. Example: `date:Y-m-d`. Default: ` `.
```php
$decrypted = Cast::cast($value, 'encrypted');
$encrypted = Cast::castDb($value, 'encrypted');
$decrypted = Cast::castJson($value, 'encrypted');
```
- **enum** - Example: `MyEnum::class`. Default: ` `.
```php
enum MyEnum : int
{
    case foo = 1;
    case bar = 2;
}

Cast::cast(1, 'enum', MyEnum::class); // MyEnum::foo
Cast::castDb(MyEnum::foo, 'enum', MyEnum::class); // 1
Cast::castJson(MyEnum::foo, 'enum', MyEnum::class); // 1
```
It can be a instance of `\Illuminate\Contracts\Support\Arrayable` or `\Illuminate\Contracts\Support\Jsonable`.
```php
use Illuminate\Contracts\Support\Arrayable;

enum MyEnum : string implements Arrayable
{
    case foo = 1;
    case bar = 2;

    public function description() : string
    {
        return match ($this) {
            self::foo => 'foo description',
            self::bar => 'bar description',
        };
    }

    public function toArray()
    {
        return [
            'name' => $this->name,
            'value' => $this->value,
            'description' => $this->description(),
        ];
    }
}

Cast::cast(1, 'enum', MyEnum::class); // MyEnum::foo
Cast::castDb(MyEnum::foo, 'enum', MyEnum::class); // 1
Cast::castJson(MyEnum::foo, 'enum', MyEnum::class);
// [
//      'name' => 'foo',
//      'value' => 1,
//      'description' => 'foo description'
// ]
```

## Custom types

To create a custom type just implements the contract `\JnJairo\Laravel\Cast\Contracts\Type`.

```php
class CustomType implements \JnJairo\Laravel\Cast\Contracts\Type
{
    // ...
}
```

Publish the configuration to `config/cast.php`.

```bash
php artisan vendor:publish --provider=JnJairo\\Laravel\\Cast\\CastServiceProvider
```

Set the new type in the configuration.

```php
// config/cast.php

return [
    'types' => [
        'custom_type' => CustomType::class,
    ],
];
```

And the custom type will be available.

```php
Cast::cast('foo', 'custom_type');
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/jn-jairo/laravel-cast/compare/v1.0.6...master)

## [v1.0.6 (2022-04-19)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.5...v1.0.6)

### Added
- Laravel 9 support
- `enum` type

## [v1.0.5 (2021-09-01)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.4...v1.0.5)

### Added
- PHP 8 support

## [v1.0.4 (2020-11-27)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.3...v1.0.4)

### Added
- `encrypted` type

## [v1.0.3 (2020-09-18)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.2...v1.0.3)

### Added
- Laravel 8 support

## [v1.0.2 (2020-08-12)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.1...v1.0.2)

### Added
- Travis PHP 7.4
- `ramsey/uuid` version `^4.0`

### Changed
- `thecodingmachine/safe` version `^1.1`

## [v1.0.1 (2020-03-08)](https://github.com/jn-jairo/laravel-cast/compare/v1.0.0...v1.0.1)

### Added
- Laravel 7 support

## [v1.0.0 (2019-12-26)](https://github.com/jn-jairo/laravel-cast/compare/v0.0.3...v1.0.0)

### Fixed
- Datetime from Carbon with wrong format

## [v0.0.3 (2019-11-30)](https://github.com/jn-jairo/laravel-cast/compare/v0.0.2...v0.0.3)

### Added
- PHPDoc in facade

## [v0.0.2 (2019-09-14)](https://github.com/jn-jairo/laravel-cast/compare/v0.0.1...v0.0.2)

### Added
- Laravel 6 support
- Test using [orchestral/testbench](https://github.com/orchestral/testbench)

### Changed
- Minimal PHP version 7.2
- Minimal Laravel version 5.8

## [v0.0.1 (2019-09-01)](https://github.com/jn-jairo/laravel-cast/commit/9a09ab219c984878ddfa124259e423d6b39dba48)
- Cast
    - `int`, `integer`
    - `float`, `real`, `double`
    - `decimal`
    - `bool`, `boolean`
    - `date`
    - `datetime`
    - `timestamp`
    - `json`
    - `array`
    - `object`
    - `collection`
    - `string`, `text`
    - `uuid`


config/cast.php
<?php

return [
    'types' => [
        'string' => \JnJairo\Laravel\Cast\Types\StringType::class,
        'text' => \JnJairo\Laravel\Cast\Types\StringType::class,

        'int' => \JnJairo\Laravel\Cast\Types\IntegerType::class,
        'integer' => \JnJairo\Laravel\Cast\Types\IntegerType::class,

        'float' => \JnJairo\Laravel\Cast\Types\FloatType::class,
        'real' => \JnJairo\Laravel\Cast\Types\FloatType::class,
        'double' => \JnJairo\Laravel\Cast\Types\FloatType::class,

        // The decimal type uses the https://php-decimal.io extension,
        // to use this type run `composer require php-decimal/php-decimal:^1.1`
        // and install the decimal extension.
        'decimal' => \JnJairo\Laravel\Cast\Types\DecimalType::class,
        /*
        'decimal' => [
            'class' => \JnJairo\Laravel\Cast\Types\DecimalType::class,
            'config' => [
                // Example using format parameter:
                // Cast::cast($value, 'decimal', '10:2|half_up');
                // Cast::cast($value, 'decimal', '10:2');
                // Cast::cast($value, 'decimal', '2');
                // Cast::cast($value, 'decimal', 'half_up');
                'precision' => 28,
                'places' => 2,
                'round_mode' => 'half_up', // up|down|ceiling|floor|half_up|half_down|half_even|half_odd|truncate
            ],
        ],
         */

        'bool' => \JnJairo\Laravel\Cast\Types\BooleanType::class,
        'boolean' => \JnJairo\Laravel\Cast\Types\BooleanType::class,

        // datetime, date and timestamp accept the same configuration parameters
        'datetime' => \JnJairo\Laravel\Cast\Types\DateTimeType::class,
        /*
        'datetime' => [
            'class' => \JnJairo\Laravel\Cast\Types\DateTimeType::class,
            'config' => [
                // Example using format parameter:
                // Cast::cast($value, 'datetime', 'Y-m-d H:i:s');
                'format' => 'Y-m-d H:i:s',
            ],
        ],
         */
        'date' => \JnJairo\Laravel\Cast\Types\DateType::class,
        'timestamp' => \JnJairo\Laravel\Cast\Types\TimestampType::class,

        'json' => \JnJairo\Laravel\Cast\Types\JsonType::class,
        'array' => \JnJairo\Laravel\Cast\Types\ArrayType::class,
        'object' => \JnJairo\Laravel\Cast\Types\ObjectType::class,
        'collection' => \JnJairo\Laravel\Cast\Types\CollectionType::class,

        // To use ordered UUID type run `composer require moontoast/math:^1.1`
        'uuid' => \JnJairo\Laravel\Cast\Types\UuidType::class,
        /*
        'uuid' => [
            'class' => \JnJairo\Laravel\Cast\Types\UuidType::class,
            'config' => [
                // Example using format parameter:
                // Cast::cast($value, 'uuid', 'uuid1');
                'format' => 'uuid4', // uuid1|uuid4|ordered
            ],
        ],
         */

        'encrypted' => \JnJairo\Laravel\Cast\Types\EncryptedType::class,

        'enum' => \JnJairo\Laravel\Cast\Types\EnumType::class,
    ],
];
